File: April-18/13-04-18/java/02-Super-two-letter-string.java

import java.util.*;

class TestClass {
    static int MOD = (int)1e9+7;
    
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int t = input.nextInt();
        while(t-- > 0)
        {
            int n, p;
            n = input.nextInt();
            p = input.nextInt();
            long dp[] = new long[n+1];
            dp[0] = 1;
            
            for(int i = 1; i <= n; i++)
            {
                int j = i - 1;
                while(j >= 0 && i-j <= p)
                {
                    dp[i] += dp[j];
                    dp[i] %= MOD;
                    j--;
                }
            }
            
            System.out.println(dp[n]);
        }
    }
}

File: April-18/13-04-18/python/02-Super-two-letter-string.py
MOD = (10**9) + 7

for _ in range(int(input())):
    n, p = map(int,input().split())
    dp = [0]*(n+1)
    dp[0] = 1
    
    for i in range(1, n+1):
        j = i-1
        while(j >= 0 and i-j <= p):
            dp[i] += dp[j]
            j -= 1
    print(dp[n] % MOD)
    
    
